### PR TITLE
Relationship properties in Kernel API

### PR DESCRIPTION
**Pull request by pontusmelke**
- Implement mutating property operations on relationships in the kernel API, i.e. `relationshipSetProperty` and `relationshipRemoveProperty`.
- Update Core API to use kernel API for all methods in `RelationshipProxy`
- Update interpreted runtime to use kernel API for all property access on relationships (and some operations that had been overlooked in previous work)
- Update compiled runtime to use kernel API for all property access on relationships.

### CODE REVIEW

File: community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java [hunk 1]
```diff
@@ -358,6 +361,247 @@ public void shouldSeeNewRelationshipPropertyInTransaction() throws Exception
     }
 
 
+
+    //start here
```
Comment (fickludd) [inline community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java]: `//start here`?

File: community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/ExplicitIndexCursorWritesTestBase.java [hunk 2]
```diff
@@ -265,7 +264,7 @@ public void shouldCreateExplicitIndexTwice() throws Exception
     }
 
     //TODO unignore when we support relationship creation.
```
Comment (fickludd) [inline community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/ExplicitIndexCursorWritesTestBase.java]: Remove the TODO as well?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java [hunk 3]
```diff
@@ -412,24 +456,28 @@ public void setProperty( String key, Object value )
             throw new IllegalStateException( "Auto indexing encountered a failure while setting property: "
                                              + e.getMessage(), e );
         }
+        catch ( KernelException e )
+        {
+            throw new ConstraintViolationException( e.getMessage(), e );
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java]: This is strange, I don't think we can get constraint violation for relationships when setting a property? When removing I guess we could violate a existence constraint, but that's anyway not checked until commit. 

In general it feels like here are way more catch clauses than necessary...

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala [hunk 4]
```diff
@@ -591,46 +588,63 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
 
     override def delete(id: Long) {
       try {
-        transactionalContext.statement.dataWriteOperations().relationshipDelete(id)
+        writes().relationshipDelete(id)
       } catch {
         case _: exceptions.EntityNotFoundException => // node has been deleted by another transaction, oh well...
       }
```
Comment (fickludd) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala]: In the Kernel API relationshipDelete is idempotent I think, so we can remove the try-catch as well.

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala [hunk 5]
```diff
@@ -642,25 +656,36 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
...
-          relationship = entityAccessor.newRelationshipProxy(relationshipId, startNodeId, typeId, endNodeId)
-        }
-      })
-      Option(fromRelationshipProxy(relationship))
-    } catch {
-      case _: exceptions.EntityNotFoundException => None
-    }
+    override def getByIdIfExists(id: Long): Option[RelationshipValue] =
+      if (reads().relationshipExists(id))
+        Some(fromRelationshipProxy(entityAccessor.newRelationshipProxy(id)))
```
Comment (fickludd) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala]: Here I preferred the previous impl which cached the start/end/type of the relationship. That stuff is going to be in L1 cache anyway (given current store layout) so we might as well return it.

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala [hunk 6]
```diff
@@ -674,24 +699,17 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
...
+    override def acquireExclusiveLock(obj: Long): Unit =
       transactionalContext.statement.readOperations().acquireExclusive(ResourceTypes.RELATIONSHIP, obj)
 
-    override def releaseExclusiveLock(obj: Long) =
+    override def releaseExclusiveLock(obj: Long): Unit =
       transactionalContext.statement.readOperations().acquireExclusive(ResourceTypes.RELATIONSHIP, obj)
 
-    override def exists(id: Long): Boolean = {
-      try {
-        transactionalContext.statement.readOperations().relationshipVisit(id, NoopVisitor)
```
Comment (fickludd) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:717]: Now we can also remove the NoopVisitor!

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala [hunk 7]
```diff
@@ -105,9 +105,9 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
   //We cannot assign to value because of periodic commit
   private def writes() = transactionalContext.dataWrite
   private def reads() = transactionalContext.dataRead
-  private val nodeCursor = allocateAndTraceNodeCursor()
-  private val relationshipScanCursor = allocateAndTraceRelationshipScanCursor()
-  private val propertyCursor = allocateAndTracePropertyCursor()
+  private lazy val nodeCursor = allocateAndTraceNodeCursor()
+  private lazy val relationshipScanCursor = allocateAndTraceRelationshipScanCursor()
+  private lazy val propertyCursor = allocateAndTracePropertyCursor()
```
Comment (fickludd) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala]: I'm curious: does this have a measurable effect, or why did you do it?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java [hunk 8]
```diff
@@ -155,6 +153,11 @@ public void delete()
             throw new IllegalStateException( "Auto indexing encountered a failure while deleting the relationship: "
                                              + e.getMessage(), e );
         }
+        catch ( EntityNotFoundException e )
+        {
+            throw new NotFoundException( "Unable to delete relationship[" +
+                                         getId() + "] since it is already deleted." );
+        }
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java]: I don't think we need this catch - this is handled by the boolean return value.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 9]
```diff
@@ -119,6 +119,11 @@ public void delete()
             throw new IllegalStateException( "Auto indexing encountered a failure while deleting the node: "
                                              + e.getMessage(), e );
         }
+        catch ( EntityNotFoundException e )
+        {
+            throw new NotFoundException( "Unable to delete node[" +
+                                         getId() + "] since it is already deleted." );
+        }
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: This should be handled by the `nodeDelete` return value, so we don't need this catch.

**Review by fickludd — CHANGES_REQUESTED** [review]
In general really good, well done on dragging this across all user and everywhere.

Some minor non-critical points, and there seems to be missing a locking tests for `relationshipDeleteProperty`.

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala (hunk 6 above)
Comment (pontusmelke) [reply community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:717]: done

**Review by fickludd — APPROVED** [review]

File: community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java [hunk 10]
```diff
@@ -312,6 +314,50 @@ public void shouldTraverseDenseNodeViaGroupsWithDetachedReferences() throws Exce
...
+
+                relationship.properties( property );
+                assertTrue( property.next() );
+                //First property
+                assertEquals( prop1, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "hello" ) );
+                //second property
+                assertTrue( property.next() );
+                assertEquals( prop2, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "world" ) );
```
Comment (chrisvest) [inline community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java]: This relies on the order in which the properties comes back from the cursor. I think that's over-specifying.

File: community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java [hunk 11]
```diff
@@ -358,6 +361,247 @@ public void shouldSeeNewRelationshipPropertyInTransaction() throws Exception
...
+
+                //property 2, start with tx state
+                assertTrue( property.next() );
+                assertEquals( propToken2, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "world" ) );
+
+                //property 1, from disk
+                assertTrue( property.next() );
+                assertEquals( propToken1, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "hello" ) );
```
Comment (chrisvest) [inline community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java]: Does this imply that the properties contained in the transaction state should always be siphoned out before the persistent state?

File: community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java [hunk 12]
```diff
@@ -358,6 +361,247 @@ public void shouldSeeNewRelationshipPropertyInTransaction() throws Exception
...
+
+        try ( org.neo4j.graphdb.Transaction ignored = graphDb.beginTx() )
+        {
+            assertThat(
+                    graphDb.getRelationshipById( relationshipId ).getProperty( propKey ), equalTo( "world" ) );
+        }
+    }
+
+    @Test
+    public void shouldSeeRemovedPropertyInTransaction() throws Exception
```
Comment (chrisvest) [inline community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java]: Or should _not_ see?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Operations.java [hunk 13]
```diff
@@ -459,10 +459,21 @@ public Value relationshipSetProperty( long relationship, int propertyKey, Value
...
     {
+        acquireExclusiveRelationshipLock( relationship );
         ktx.assertOpen();
-        throw new UnsupportedOperationException();
+        singleRelationship( relationship );
+        Value existingValue = readRelationshipProperty( propertyKey );
+
+        if ( existingValue != NO_VALUE )
+        {
+            //no existing value, we just add it
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Operations.java]: This comment look out of place.

**Review by chrisvest — APPROVED** [review]
Couple of comments, but overall it looks good.

**Pull request merged**